src/platforms/atomic-kms/server/kms/quirks.cpp

```diff
@@ -56,100 +60,221 @@ class mga::Quirks::Impl
+            {
+                return ::gbm_surface_has_free_buffers(gbm_surface);
+            }
+        };
+
+        auto const driver = mgc::get_device_driver(device.parent().get());
+        auto const devnode = device.devnode();
+        mir::log_debug("Quirks(gbm-create-surface-flags + gbm-surface-has-free-buffers): checking device with devnode: %s, driver %s", devnode, driver);
+
+        auto const devnode_or_driver =
+            [](auto devnode, auto driver, auto devnodes, auto drivers) -> std::optional<std::string>
+        {
+            if (devnodes.contains(devnode))
+                return devnodes.at(devnode);
+
+            if (drivers.contains(driver))
+                return drivers.at(driver);
+
+            return std::nullopt;
+        };
+
+        auto create_surface_impl_name =
+            devnode_or_driver(devnode, driver, gbm_create_surface_flags.devnodes, gbm_create_surface_flags.drivers)
+                .transform(
+                    [](auto impl_name)
+                    {
+                        mir::log_debug(
+                            "Quirks(gbm-create-surface-flags): forcing %s implementation", impl_name.c_str());
+                        return impl_name;
+                    })
+                .or_else(
+                    [&driver] -> std::optional<std::string>
+
+                    {
+                        mir::log_debug(
+                            "Quirks(gbm-create-surface-flags): using default implementation for %s driver", driver);
+                        // Not specified
+                        return driver;
+                    })
+                .value();
+
+        auto create_surface_impl = [&]() -> std::unique_ptr<GbmQuirks::CreateSurfaceFlagsQuirk>
+        {
+            if (create_surface_impl_name == "no-flags" || create_surface_impl_name == "nvidia")
+                return std::make_unique<GbmCreateSurfaceNoFlags>();
+
+            return std::make_unique<DefaultGbmCreateSurfaceFlags>();
+        }();
+
+        auto surface_has_free_buffers_impl_name =
+            devnode_or_driver(
+                devnode, driver, gbm_surface_has_free_buffers.devnodes, gbm_surface_has_free_buffers.drivers)
+                .transform(
+                    [](auto impl_name)
+                    {
+                        mir::log_debug(
+                            "Quirks(gbm-surface-has-free-buffers): forcing %s implementation", impl_name.c_str());
+                        return impl_name;
+                    })
+                .or_else(
+                    [&driver] -> std::optional<std::string>
+
+                    {
+                        mir::log_debug(
+                            "Quirks(gbm-surface-has-free-buffers): using default implementation for %s driver", driver);
+                        // Not specified
+                        return driver;
+                    })
+                .value();
```

Comment:
I can guess that TICS is upset because the (temporary) `optional<>` returned by `.or_else()` is not being checked when `.value()` is used. It isn't obvious why we create that optional simply to fetch a value we know is there:

```suggestion
                .value_or(
                    [&driver]

                    {
                        mir::log_debug(
                            "Quirks(gbm-surface-has-free-buffers): using default implementation for %s driver", driver);
                        // Not specified
                        return driver;
                    });
```

Reply:
Actually, that's suggestion is not right: `.value_or()` doesn't evaluate a functor...